server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationConstraints.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright OpenSearch Contributors.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.cluster.routing.allocation;
+
+import org.opensearch.cluster.routing.allocation.allocator.BalancedShardsAllocator;
+import org.opensearch.common.settings.Setting;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Predicate;
+
+/**
+ * Allocation constraints specify conditions which, if breached, reduce the priority of a node
+ * for receiving shard allocations.
+ */
+public class AllocationConstraints {
```

Comment:
If you think that these constraints are optional and don't apply for all use-cases, we could add an AllocationConstraint plugin in opendistro and create a framework in OpenSearch for the same. That way, you could write your own specific constraint logic. Adding an optional setting in Opensearch for disabling the constraint mode per constraint does not seem right to me.

Reply:
This is an interesting topic. We can argue this for most settings and features - keep them in a plugin where they are independently extended. But plugins come with a non-trivial cost. Users will have to separately download and set it up on their clusters. Who is going to maintain that plugin? If users want their own specific constraints, they can always contribute to OpenSearch, which will be similar to contributing upstream to the plugin repo.

In my opinion, we should make something into a plugin, if it is big enough to warrant its own independent development cycle. I don't think that is true for Allocation Constraints. It is an override on a very small (albeit important) part of the BalancedShardsAllocator, the weight fn.

Keeping it in OpenSearch allows users to get this functionality right out of the box. In my past experience with operating ES clusters, _indexing hotspots_ were a very common problem, with users trying to work past it using suboptimal shards-per-node settings. It would be good if we can get this with the default engine.

Reply:
Given this falls back to the original weight function, it seems to solve almost all use-cases. We could keep it in the engine, but this looks like a good candidate for allowing plugins to add more constraints. We can think more on it in future.